src/tensorwaves/model.py

```diff
@@ -77,6 +143,135 @@
         self.__parameters.update(new_parameters)
 
 
+class _ConstantSubExpressionSympyModel(Model):
+    """Implements a performance optimized sympy based model.
+
+    Based on which symbols of the sympy expression are declared
+    """
+
+    # pylint: disable=too-many-instance-attributes
+    def __init__(
+        self,
+        expression: sp.Expr,
+        parameters: Dict[sp.Symbol, Union[float, complex]],
+        fix_inputs: DataSample,
+    ) -> None:
+        self.__fix_inputs = fix_inputs
+        self.__constant_symbols = set(self.__fix_inputs)
+        self.__constant_sub_expressions: Dict[sp.Symbol, sp.Expr] = {}
+        self.__find_constant_subexpressions(expression)
+        self.__expression = self.__replace_constant_sub_expressions(expression)
+        self.__not_fixed_parameters = {
+            k: v
+            for k, v in parameters.items()
+            if k.name not in self.__constant_symbols
+        }
+        self.__not_fixed_variables: FrozenSet[sp.Symbol] = frozenset(
+            s
+            for s in self.__expression.free_symbols
+            if s.name not in self.parameters
+            and s.name not in self.__constant_symbols
+            and s not in self.__constant_sub_expressions
+        )
+        self.__argument_order = tuple(self.__not_fixed_variables) + tuple(
+            self.__not_fixed_parameters
+        )
+
+    def __find_constant_subexpressions(self, expr: sp.Expr) -> bool:
+        if not expr.args:
+            if (
+                isinstance(expr, sp.Symbol)
+                and expr.name not in self.__constant_symbols
+            ):
+                return False
+            return True
+
+        is_constant = True
+        temp_constant_sub_expression = []
+        for arg in expr.args:
+            if self.__find_constant_subexpressions(arg):
+                if arg.args:
+                    temp_constant_sub_expression.append(arg)
+            else:
+                is_constant = False
+
+        if not is_constant:
+            for sub_expr in temp_constant_sub_expression:
+                placeholder = sp.Symbol(f"cached[{str(sub_expr)}]")
+                self.__constant_sub_expressions[placeholder] = sub_expr
+        return is_constant
+
+    def __replace_constant_sub_expressions(
+        self, expression: sp.Expr
+    ) -> sp.Expr:
+        new_expression = copy.deepcopy(expression)
```

Comment:
I'm surprised this `deepcopy` is necessary. Does `xreplace`/`subs` modify the expression?

Reply:
Good question. I'm actually not sure on the internals of the sympy expression, but the original expression not the be modified